shakenfist/db.py

```diff
@@ -146,47 +162,83 @@ def persist_ipmanager(network_uuid, data):
 def delete_ipmanager(network_uuid):
     etcd.delete('ipmanager', None, uuid)
 
+#####################################################################
+# Instances
+#####################################################################
+
+# Theorem: we never write to the instance object, except for when we
+# create or delete it. Anything which changes value during the lifetime
+# of the instance (state, power_state, etc) is stored in a set of
+# secondary values in etcd that we never cache. This is version 2 of
+# the instance object.
+
+
+def create_instance(instance_uuid, metadata):
+    set_instance_attribute(instance_uuid, 'state', {'state': 'initial'})
+    etcd.put('instance', None, instance_uuid, metadata)
+
+
+def get_instance(instance_uuid, external_view=False):
+    # NOTE(mikal): we don't do upgrades inflight. They are assumed to have been
+    # done as part of the upgrade process.
+    i = etcd.get('instance', None, instance_uuid)
+    if i.get('version') != 2:
+        raise exceptions.BadObjectVersion(
+            'Unknown version - Instance: %s', i)
 
-def get_instance(instance_uuid):
-    return etcd.get('instance', None, instance_uuid)
+    # If this is an external view, then mix back in attributes that users expect
+    if external_view:
```

Comment:
These could be retrieved via a property function.

Reply:
Is that true if we're returning a dict here not an object? I wanted to have these be explicit so we don't rely on this behaviour in internal code, its just here so API callers get the data back in the right format.

Reply:
I was hoping to remove system logic from the DB interface layer. All instance logic could (should) be in the Instance class. Leaving DB to the be shim that gathers/persists data.